rector.php

```diff
@@ -0,0 +1,52 @@
+<?php
+
+/*
+ * This file is part of the Symfony MakerBundle package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+use Rector\Config\RectorConfig;
+use Rector\Php74\Rector\Closure\ClosureToArrowFunctionRector;
+use Rector\Php74\Rector\Property\RestoreDefaultNullToNullableTypePropertyRector;
+use Rector\Php80\Rector\Catch_\RemoveUnusedVariableInCatchRector;
+use Rector\Php80\Rector\Class_\ClassPropertyAssignToConstructorPromotionRector;
+use Rector\Php80\Rector\Identical\StrEndsWithRector;
+use Rector\Php80\Rector\NotIdentical\StrContainsRector;
+use Rector\Php80\Rector\Switch_\ChangeSwitchToMatchRector;
+use Rector\Php81\Rector\Array_\FirstClassCallableRector;
+use Rector\Php81\Rector\FuncCall\NullToStrictStringFuncCallArgRector;
+use Rector\Php81\Rector\Property\ReadOnlyPropertyRector;
+use Rector\Set\ValueObject\LevelSetList;
+use Rector\TypeDeclaration\Rector\ClassMethod\ReturnNeverTypeRector;
+
+return static function (RectorConfig $config): void {
+    $config->paths([
+        __DIR__.'/src',
+    ]);
+
+    $config->rules([
+        // Remove when LevelSet is fully enabled
+        ClassPropertyAssignToConstructorPromotionRector::class,
+        ClosureToArrowFunctionRector::class,
+        RemoveUnusedVariableInCatchRector::class,
+        FirstClassCallableRector::class,
+        StrEndsWithRector::class,
+//        StrContainsRector::class,
+        RestoreDefaultNullToNullableTypePropertyRector::class,
+        ReturnNeverTypeRector::class,
+        ChangeSwitchToMatchRector::class,
+    ]);
+
+    $config->sets([
+//        LevelSetList::UP_TO_PHP_81,
+    ]);
+
+    $config->skip([
+        NullToStrictStringFuncCallArgRector::class,
+        ReadOnlyPropertyRector::class,
+    ]);
+};
```

Comment:
Do we actually need rector in the project? Or can you just use it locally this one time, and going forward, we rely on something like php-cs-fixer or phpstan?

Reply:
I think having it in the project will be beneficial in the long run. `rector.php` is just the config file (unfortunately rector doesnt pickup on `rector.dist.php` by default) - similar to `.php-cs-fixer.dist.php`. Rector itself lives in `tools/rector` like `tools/php-cs-fixer` does - it doesnt cause composer dependency headaches.

The `php-cs-fixer` rulesets for PHP are pretty lacking in terms of upgrading the codebase to a language level. Rector is for more robust in this aspect. The same holds true for `psalm` - its "auto fix" is limited in scope leading to more manual work. && `phpstan` auto fixing appears to be a paid feature.  I would however like to get `phpstan` working against the codebase (in a future pr). I've seen a handful of codesmells that a static analyser would have flagged for us long ago..

having the config file in the codebase at a minimum will be a leg up if/when we enable this in CI as a sniff tests..

Reply:
Unless having rector in the project is standard elsewhere (and I'm not aware of anything), I'm 👎  on including it. The process should be:

A) A one-time rector usage to fix the project
B) Some tool to fail in CI if we're violating standards. I don't care as much about the autofixing aspect. If php-cs-fixer is missing a check - like for property promotion - I think we just track that manually in PR's.

In short, I don't want to do anything non-standard or extra-forward in this repo. I feel like these files will get forgotten and just get out of date. Sorry :)